### DISCUSSION TITLE
Custom response format for validation error

### DISCUSSION THREAD

**Discussion by LKay** [fastapi/fastapi#8298] [Questions]
How can I reformat default response for validation errors? I find the default formatting too verbose for my app, and wanted to return errors in different structure. 

I thought of adding error handler explicitly for `pydantic.ValidationError` but this is ignored as the error is caught before that and handled internally by FastAPI:

https://github.com/tiangolo/fastapi/blob/4a1648b04e74b5943f3ffadf92737f7d9c8201af/fastapi/routing.py#L68-L72

Is this possible to override default formatter and handle `ValidationError` on my own? I tried to search through issues and docs but couldn't find anything related.

**Comment by tiangolo** [comment]
You could add a custom handler for 422 errors: https://fastapi.tiangolo.com/tutorial/handling-errors/#installing-custom-handlers

And extract the original message from the `HTTPExcpetion.detail`.

Let me know if that works for you.

Otherwise, we can think about refactoring how `ValidationError`s are handled to allow custom handlers (without passing through the `HTTPException` error).

**Comment by LKay** [comment]
Some way of providing custom handler for validation errors would be nice. Currently I added handler for 422 errors but it doesn’t solve the issue entirely. I could reformat the dict, however it would make more sense to just get the original exception somehow and prices that. The other thing is, adding handler for 422 status doesn’t overwrite APIdocs for that status even if I added custom 422 definition in `responses` for `@api.route()`. Is there any way to override the default until some more flexible validation error handler is available?

**Comment by tiangolo** [comment]
First, about having custom validation handlers, I'll take it as a feature request.

For overriding 422 errors, I would need to check it. I think that might be another feature request/issue too.

**Comment by tiangolo** [comment]
This is supported now in the latest versions. :tada:

You can now add handlers for `RequestValidationError` and get the full Pydantic `ValidationError`: https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions

Make sure to update your FastAPI version. The current latest is `0.27.0`.

I'll close this issue now, but feel free to add more comments or create new issues.

**Comment by Olegt0rr** [comment]
> This is supported now in the latest versions. 🎉
> 
> You can now add handlers for `RequestValidationError` and get the full Pydantic `ValidationError`: https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions
> 
> Make sure to update your FastAPI version. The current latest is `0.27.0`.
> 
> I'll close this issue now, but feel free to add more comments or create new issues.

Docs builders don't know about exception overriding.
Please add an example [this page](https://fastapi.tiangolo.com/tutorial/handling-errors/#override-request-validation-exceptions). 

### Screenshot
```python
@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    return PlainTextResponse(str(exc), status_code=422)
```
![image](https://user-images.githubusercontent.com/25399456/102081153-d0ecfd80-3e20-11eb-9745-18b445f0dd1b.png)

**Comment by jdavis-st** [comment]
This needs to be reopened. There's no sense in customizing an error response message if it can't be picked up by the API docs to inform users.

**Comment by gistart** [comment]
Is there a way to hide 422 error code from documentation?

**Comment by wimvc1** [comment]
> This needs to be reopened. There's no sense in customizing an error response message if it can't be picked up by the API docs to inform users.

For future reference:
If you provide a custom model, the OpenAPI specs will pick them up.

```
@app.post(
    "/cookies",
    status_code=201,
    responses={
        201: {"model": MyCustomOutputModel},
        422: {"model": MyValidationModel},
        500: {"model": UnexpectedServerError},
    },
    include_in_schema=True,
)
```

**Comment by Raphasha27** [comment]
Override FastAPI's default validation error handler with a custom exception handler for `RequestValidationError`:

```python
from fastapi import FastAPI, Request
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse

app = FastAPI()

@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request: Request, exc: RequestValidationError):
    errors = []
    for error in exc.errors():
        errors.append({
            "field": ".".join(str(l) for l in error["loc"]),
            "message": error["msg"],
            "code": error["type"],
        })
    return JSONResponse(
        status_code=422,
        content={"status": "error", "errors": errors},
    )
```

To also handle 422 errors from Starlette's internal validation (like path params/multipart), add a handler for `StarletteHTTPException` with status 422:

```python
from fastapi.exceptions import StarletteHTTPException

@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request: Request, exc: StarletteHTTPException):
    if exc.status_code == 422:
        return JSONResponse(
            status_code=422,
            content={"status": "error", "message": str(exc.detail)},
        )
    return JSONResponse(status_code=exc.status_code, content={"detail": exc.detail})
```

This gives you full control over the response structure without touching the default validation logic.
